src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,118 @@
+
         self.expected_message = expected_message
+        self.dev_message = dev_message
 
-    def __enter__(self):
-        pass
+    def _check_dev_message(self, exc: ContractLogicError):
+        """
+        Attempts to extract a dev-message from the contract source code by inspecting what
+        instruction(s) led to a transaction revert.
 
-    def __exit__(self, exc_type: Type, exc_value: Exception, traceback) -> bool:
-        if exc_type is None:
-            raise AssertionError("Transaction did not revert.")
+        TODO: Document exception-raising cases
+        """
+        if exc.txn is None:
+            raise AssertionError("Could not fetch transaction information to check dev message.")
 
-        if not isinstance(exc_value, ContractLogicError):
+        receipt = exc.txn.provider.get_receipt(
+            exc.txn.txn_hash,
+            raise_on_fail=False,
+        )
+
+        try:
+            contract = ape.Contract(receipt.receiver)
+        except ValueError as exc:
             raise AssertionError(
-                f"Transaction did not revert.\n"
-                f"However, an exception of type {type(exc_value)} occurred: {exc_value}"
-            ) from exc_value
+                f"Could not fetch contract at {receipt.receiver} to check dev message"
+            ) from exc
+
+        if contract.contract_type.sourcemap is None:
+            raise AssertionError("Could not fetch contract source to check dev message")
+
+        trace = deque(exc.trace)
+
+        if trace[-1].op != "REVERT":
+            raise AssertionError("Transaction trace did not end with REVERT")
+
+        pc = None
+
+        # To find a suitable line for inspecting dev messages, we must start at the revert and work
+        # our way backwards. If the last frame's PC is in the PC map, the offending line is very
+        # likely a 'raise' statement.
+        if str(trace[-1].pc) in contract.contract_type.pcmap:
+            pc = str(trace[-1].pc)
+
+        # Otherwise we must traverse the trace backwards until we find our first suitable candidate.
+        # It should be noted that when an error happens, typically the contract will execute:
+        #
+        #   <...>
+        #   JUMPI    <- jump to revert routine on a failing condition
+        #
+        #   <...>
+        #   JUMPDEST <- beginning of revert routine
+        #   PUSH1
+        #   DUP1
+        #   REVERT
+        #
+        # So we must step backwards until we find the initial instructions that caused the failing
+        # condition. This will not be the JUMPI or even an arithmetic instruction like DIV, but
+        # rather the starting instruction for a given line.
+        else:
+            while len(trace) > 0:
+                frame = trace.pop()
+                if str(frame.pc) in contract.contract_type.pcmap:
+                    pc = str(frame.pc)
+
+                    break
```

Comment:
@unparalleled-js @fubuloubu @NotPeopling2day 

So I noted this in the Discord as well, but the `pc_pos_map` seems to align with the trace for the _start_ of a line. In the case of my zero-division example I was trying, the instruction I was actually finding in the PC map was `DUP1`

I originally was looking for the last `JUMPI` or so but that won't align with the PC map so I decided on this instead: stepping backwards to the first matching line.

Thoughts, and are there any caveats there may be from this?

I think for transparency's sake if we can't find the dev message, we should at least print out the candidate line we looked at (if we had such a line)